solrmonitor/zkdispatcher_test.go

```diff
@@ -0,0 +1,214 @@
+	// itself
+	maxNumEvents int32
+	// if non-nil, mark this as done at the start of first callback (to indicate that the
+	// callback is ready to start)
+	ready *sync.WaitGroup
+	// if non-nil, wait on this before starting first callback (to coordinate the start of
+	// multiple callbacks)
+	run *sync.WaitGroup
+}
+
+func (cb *testCallback) Handle(e zk.Event) <-chan zk.Event {
+	// track the maximum number of concurrent executions
+	num := atomic.AddInt32(&cb.state.numConcurrentExecutions, 1)
+	defer atomic.AddInt32(&cb.state.numConcurrentExecutions, -1)
+
+	// if these wait groups are non-nil, we are coordinating the start of handling
+	if cb.ready != nil {
+		cb.ready.Done()
+		cb.ready = nil
+	}
+	if cb.run != nil {
+		cb.run.Wait()
+		cb.run = nil
+	}
+
+	for {
+		max := atomic.LoadInt32(&cb.state.maxConcurrentExecutions)
+		if num <= max {
+			break
+		}
+		if num > max &&
+			atomic.CompareAndSwapInt32(&cb.state.maxConcurrentExecutions, max, num) {
+			break
+		}
+	}
+
+	numEvents := atomic.AddInt32(&cb.numEvents, 1)
+	if cb.state.assumeEventOrder && e.Type != zk.EventType(numEvents) {
+		cb.state.t.Errorf("Expecting event %d, got %d", numEvents, int(e.Type))
+	}
+
+	if numEvents >= cb.maxNumEvents {
+		cb.state.done.Done()
+		return nil
+	} else {
+		ch := make(chan zk.Event, 1)
+		ch <- cb.state.gen.newEvent()
+		return ch
+	}
+}
+
+func TestReregistration(t *testing.T) {
+	disp := NewZkDispatcher(zk.DefaultLogger)
+
+	state := testState{t: t, assumeEventOrder: true}
+	cb := state.newCallback(100)
+	state.done.Add(1)
+
+	channel := make(chan zk.Event, 1)
+	disp.WatchEvent(channel, cb)
+
+	channel <- state.gen.newEvent()
+	state.done.Wait()
+
+	if cb.numEvents != 100 {
+		t.Errorf("Should have processed 100 events but instead processed %d", cb.numEvents)
+	}
+
+	tasksShouldBecomeEmpty(t, disp)
+}
+
+func TestDispatchesToSameCallbackAreSerial(t *testing.T) {
+	disp := NewZkDispatcher(zk.DefaultLogger)
+	var channels [10]chan zk.Event
+	var run sync.WaitGroup
+	run.Add(1)
+
+	state := testState{t: t, run: &run}
+	cb := state.newCallback(10)
+
+	for i := 0; i < 10; i++ {
+		state.done.Add(1)
+		channels[i] = make(chan zk.Event, 1)
+		disp.WatchEvent(channels[i], cb)
+	}
+
+	// send an event to get them started
+	for i := 0; i < 10; i++ {
+		channels[i] <- state.gen.newEvent()
+	}
+	// give dispatch go routine(s) a chance to start
+	time.Sleep(200 * time.Millisecond)
+	// let 'em rip
+	run.Done()
+
+	// wait for them all to finish
+	state.done.Wait()
+
+	if state.maxConcurrentExecutions != 1 {
+		t.Errorf("Executions should be serial, but detected %d concurrent exceutions",
+			state.maxConcurrentExecutions)
+	}
+
+	tasksShouldBecomeEmpty(t, disp)
+}
+
+func TestDispatchesToDifferentCallbacksAreConcurrent(t *testing.T) {
+	// make sure we can run all of these concurrently
+	numRoutines := runtime.GOMAXPROCS(0)
+	if numRoutines < 30 {
+		numRoutines = runtime.GOMAXPROCS(30)
+	}
+	defer runtime.GOMAXPROCS(numRoutines)
```

Comment:
You're right, I think it is not necessary. I was temporarily thinking about go routines incorrectly, as if they were heavy-weight threads, failing to realize that this test could make forward progress and pass all test expectations even with only one worker...